### ISSUE TITLE
RNTN fit function does not work

### ISSUE THREAD

**Issue by ts335793** [deeplearning4j/deeplearning4j#225]
getValueGradient function in RNTN does not calculate gradient. Code inside if statements in lines 933 - 967 is never reached.

Fit function creates threads and does not wait for them to finish. It also does not return any thread handlers.

Possible solution:
https://github.com/ts335793/deeplearning4j/commit/617754401f34c24b30c6e631151fb20c71939b70

**Comment by agibsonccc** [comment]
Thanks for the headsup. Would you mind submitting a pull request? I can work on it a bit more there.

**Comment by Bam4d** [comment]
This issue is also causing me some troubles. I'll get a PR done ASAP when I understand whats going on better

**Comment by Bam4d** [comment]
51065  [ka.actor.default-dispatcher-26] WARN  Parallelization           - Error occurred processing data 
java.lang.IllegalStateException: Given ndarray does not have data type double
	at org.nd4j.linalg.factory.DataTypeValidation.assertDouble(DataTypeValidation.java:40) ~[nd4j-api-0.0.3.5.5.3-SNAPSHOT.jar:na]

Getting this error now... any ideas why ?

**Comment by agibsonccc** [comment]
Your data types need to line up. For now at the beginning of your program just put:

Nd4j.dtype = DataBuffer.FLOAT;

The root cause of the problem is arrays being created that are inconsistent with the data type for nd4j.

**Comment by Bam4d** [comment]
The offending line that causes the above exception is: ``` Double nodeWeight = classWeights.get(goldClass); ```

if I set dtype to FLOAT then ```nodeWeight``` is still a double and causes the same exception...

**Comment by Bam4d** [comment]
changing nodeWeight to a float and forcing various things to be floats I then start getting this error:

```
o.d.p.Parallelization - Error occurred processing data
org.jblas.exceptions.LapackException: LAPACK Java.raxpy: Parameters for x aren't valid! (n = 10000, dx.length = 500000, dxIdx = 0, incx = 5000)
	at org.jblas.JavaBlas.raxpy(JavaBlas.java:213) ~[jblas-1.2.3.jar:na]
	at org.nd4j.linalg.jblas.BlasWrapper.axpy(BlasWrapper.java:195) ~[nd4j-jblas-0.0.3.5.5.3-SNAPSHOT.jar:na]
	at org.nd4j.linalg.api.ndarray.BaseNDArray.addi(BaseNDArray.java:2556) ~[nd4j-api-0.0.3.5.5.3-SNAPSHOT.jar:na]
	at org.nd4j.linalg.api.ndarray.BaseNDArray.addi(BaseNDArray.java:2529) ~[nd4j-api-0.0.3.5.5.3-SNAPSHOT.jar:na]
	at org.nd4j.linalg.api.ndarray.BaseNDArray.add(BaseNDArray.java:2290) ~[nd4j-api-0.0.3.5.5.3-SNAPSHOT.jar:na]```

Any help would be super appreciated

**Comment by agibsonccc** [comment]
Is this with the latest nd4j master?

**Comment by Bam4d** [comment]
I'm using whatever gets set up when doing setup.sh. so If that downloads the latest nd4j then yes... But im guessing that's not right?


Also, in this pull request I'm essentially allowing multiple fit() functions to be run at once, does this make sense with the implementation of a neural network? surely the threads are just going to be overwriting the private functions within the RNTN network? How are the results of each epoch being combined to adjust the weights of the network?

How is a parallel implementation of this is even possible ?

sorry for all the questions. :)

**Comment by agibsonccc** [comment]
Ok great. And no: please keep asking. Also the parallelism part is just computing different parts of the gradient in parallel. It's equivalent to a for loop. There shouldn't be any param averaging or anything.

For any parallel stuff we do anywhere else in dl4j (spark,hadoop,..) we do parameter averaging (the same stuff google does)

**Comment by Bam4d** [comment]
OK I see the ```setParams(params.subi(gradient));``` I'm guessing that's where stuff gets stored before parameter averaging?

another question: my lookup table for words is essentially 30k ish and layersize in word2vec is about 200, some of my trees have about 1k leaf nodes... Is it possible to backprop the errors that far? would it just take a huge amount of time? (currently using jblas on 8 cores and 30g of RAM).

I'm guessing its probably unreasonable to think this would work on something that doesn't use GPU

**Comment by agibsonccc** [comment]
For an individual RNTN no param averaging of any kind should be done ever.

setParams is for an update.

Yes. It's a binary tree. For the RNTN it's based on the constituency parse of the tree.

**Comment by Bam4d** [comment]
OK I misread your comment about that it didn't do param averaging.

So the part in fit() where it is creating new threads for each tree and running all the computation in parallel for those is not a problem at all. That's good to know (means my PR makes sense).

One final question for tonight.. I've not scaled up this kinda processing stuff before.. can you point me at any resources/examples of RNTN to say what sort of clusters/hardware i should look at running this stuff on?

**Comment by agibsonccc** [comment]
Sure. You're right to use GPUs and the like. After I merge your PR I will actually integrate this in to our layer based interface. This will allow you to use spark/hadoop.

I have a question about your use case though. Are you goiung to be trying to train an RNTN on a whole paragraph? It's typically meant for phrase level extraction.

**Comment by Bam4d** [comment]
I can't go into too much detail as its very experimental, but I'm actually using it to classify information stored in abstract tree structures in XML-like data, potentially classifying nodes of the tree as well as the overall tree. So i'm not using sentence trees (might incorprate them later)

Thanks very much for your help. It seems like the algorithm has run OK over night. only a fraction of the actual examples trained out of the whole sample, but it's a start. Going to go launch a GPU box

**Comment by Bam4d** [comment]
P.S it would be super nice to get that RottenTomatoes example working on 0.0.3.3. Lots of people seem to be interested in it

**Comment by agibsonccc** [comment]
Yeah. I agree completely. I will close this. After I get convolutions in shape I'll get the RNTN in line.
